lib/Conversion/VerifToSMT/VerifToSMT.cpp

```diff
@@ -146,6 +168,230 @@ struct LogicEquivalenceCheckingOpConversion
+      return failure();
+
+    unsigned numRegs =
+        cast<IntegerAttr>(op->getAttr("num_regs")).getValue().getZExtValue();
+
+    auto initFuncTy = rewriter.getFunctionType({}, initOutputTy);
+    // Loop and init output types are necessarily the same, so just use init
+    // output types
+    auto loopFuncTy = rewriter.getFunctionType(loopInputTy, initOutputTy);
+    auto circuitFuncTy =
+        rewriter.getFunctionType(circuitInputTy, circuitOutputTy);
+
+    func::FuncOp initFuncOp, loopFuncOp, circuitFuncOp;
+
+    {
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointToEnd(
+          op->getParentOfType<ModuleOp>().getBody());
+      initFuncOp = rewriter.create<func::FuncOp>(loc, names.newName("bmc_init"),
+                                                 initFuncTy);
+      rewriter.inlineRegionBefore(op.getInit(), initFuncOp.getFunctionBody(),
+                                  initFuncOp.end());
+      loopFuncOp = rewriter.create<func::FuncOp>(loc, names.newName("bmc_loop"),
+                                                 loopFuncTy);
+      rewriter.inlineRegionBefore(op.getLoop(), loopFuncOp.getFunctionBody(),
+                                  loopFuncOp.end());
+      circuitFuncOp = rewriter.create<func::FuncOp>(
+          loc, names.newName("bmc_circuit"), circuitFuncTy);
+      rewriter.inlineRegionBefore(op.getCircuit(),
+                                  circuitFuncOp.getFunctionBody(),
+                                  circuitFuncOp.end());
+      auto funcOps = {&initFuncOp, &loopFuncOp, &circuitFuncOp};
+      // initOutputTy is the same as loop output types
+      auto outputTys = {initOutputTy, initOutputTy, circuitOutputTy};
+      for (auto [funcOp, outputTy] : llvm::zip(funcOps, outputTys)) {
+        auto operands = funcOp->getBody().front().back().getOperands();
+        rewriter.eraseOp(&funcOp->getFunctionBody().front().back());
+        rewriter.setInsertionPointToEnd(&funcOp->getBody().front());
+        SmallVector<Value> toReturn;
+        for (unsigned i = 0; i < outputTy.size(); ++i)
+          toReturn.push_back(typeConverter->materializeTargetConversion(
+              rewriter, loc, outputTy[i], operands[i]));
+        rewriter.create<func::ReturnOp>(loc, toReturn);
+      }
+    }
+
+    auto solver =
+        rewriter.create<smt::SolverOp>(loc, rewriter.getI1Type(), ValueRange{});
+    rewriter.createBlock(&solver.getBodyRegion());
+
+    // Call init func to get initial clock values
+    ValueRange initVals =
+        rewriter.create<func::CallOp>(loc, initFuncOp)->getResults();
+
+    // InputDecls order should be <circuit arguments> <state arguments>
+    // <wasViolated>
+    // Get list of clock indexes in circuit args
+    size_t initIndex = 0;
+    SmallVector<Value> inputDecls;
+    SmallVector<int> clockIndexes;
+    for (auto [curIndex, types] :
+         llvm::enumerate(llvm::zip(oldCircuitInputTy, circuitInputTy))) {
```

Comment:
Nit: Enumerate can zip at the same time
```suggestion
         llvm::enumerate(oldCircuitInputTy, circuitInputTy)) {
```

Reply:
:star_struck: